### ISSUE TITLE
V3 brainstorm

### ISSUE THREAD

**Issue by brendt** [spatie/data-transfer-object#151]
# V3 brainstorm

When this package started, it was meant to solve four problems:

- Runtime type checks for class properties
- Support for union types
- Support for array types (these are not full blown generics!)
- Named arguments

Typed properties are added in PHP 7.4, and union types and named arguments in PHP 8. So now that PHP 8 GA is around the corner, it's time to think what should and shouldn't be added in V3. Because the goal of this package was to become obsolete at one point in the future, I think it's important to critically look at our work and evaluate whether it's worth keep supporting the package for PHP 8 in its current form.

So in this post I'll write down all my thoughts, and share what I think is the right way for this package to evolve to. 

## Less is more

The more type checks supported by PHP and static analysers, the better. It means that we don't have to maintain our own runtime type checker, and that code will be more performant overall. With the addition of typed properties and union types, I doubt V3 should still implement its own type checker.

So what's left in this package if we remove type checks?

- Named arguments
- Array types
- Value casts

Let's discuss them one by one

### Named arguments

These are also obsolete since PHP 8, there's nothing more to say.

### Array types

I don't think we need to check array types at runtime anymore, for several reasons:

- Static analysis usually offers the most benefits. PhpStorm and all static analysis tools support them.
- If you really want runtime type checking, you could still create your own `DataTransferObjectCollection` class.

### Value casts

While originally not in the scope of this package, it turned out to be a rather popular feature. Currently it allows to write DTOs like this:

```php
class A extends DateTransferObject
{
    public B $child;
}

class B extends DateTransferObject
{
    public string $name;
}

$a = new A([
    'b' => [
        'name' => 'foo',
    ],
]);
```

While I personally never use this feature (I prefer to manually map raw data to my DTOs), I know that the majority of people like it.

## What's left

If we decide to drop type checks altogether (including array types), there's only one real feature left: automatic value casting. There has also been a few requests to allow more types of value casts. So what if we'd try something like this:

```php
class NewDTO extends DataTransferObject
{
    public string $name;
    
    public OtherDTO $child;
    
    #[CastWith(ComplexObjectCast::class)]
    public ComplexObject $complexObject;
    
    /** @var int[] */
    public array $listOfInts;
    
    public StringCollection $listOfStrings; 
}

class OtherDTO extends DataTransferObject
{
    public string $name;
}
```

This DTO could be constructed in several ways. First using named arguments:

```php
$dto = new NewDTO(
    name: 'test',
    child: [ // Would be casted automatically to OtherDTO
        'name' => 'other'
    ],
    complexObject: [ // Would be casted to ComplexObject using ComplexObjectCast
        'id' => 1   
    ],
    listOfInts: [1, 2], // No additional type checks
    listOfStrings: ['a', 'b'], // Would be casted automatically to StringCollection
);
```

But als still using the old array approach:

```php
$dto = new NewDTO([
    'name' => 'test',
    'child' => [ // Would be casted automatically to OtherDTO
        'name' => 'other'
    ],
    'complexObject' => [ // Would be casted to ComplexObject using ComplexObjectCast
        'id' => 1   
    ],
    'listOfInts' => [1, 2], // No additional type checks
    'listOfStrings' => ['a', 'b'], // Would be casted automatically to StringCollection
]);
```

I'm ok to use `FlexibleDataTransferObject` as the only implementation, so that additional data not present on the object is simply ignored and doesn't thrown an exception.

**Comment by mfn** [comment]
> First using named arguments:

I followed PHP8 but was not aware that you could do this, i.e. the arguments actually don't need to be featured in the constructor and using `func_get_args()` (I presume) you basically have a (slightly) shorted version to write a hash? 🤷‍♀️

**Comment by mfn** [comment]
Ok, no doesn't work, must be missing something? => https://3v4l.org/oXnjd
```php
<?php declare(strict_types=1);

class DTO {
    public function __construct() {
        var_dump(func_get_args());
    }
}

new DTO(foo: 'bar', baz: 'daz');
```
PHP 8:
```
Fatal error: Uncaught Error: Unknown named parameter $foo in /in/oXnjd:9
Stack trace:
#0 {main}
  thrown in /in/oXnjd on line 9

Process exited with code 255.
```

**Comment by brendt** [comment]
You can use this syntax:

```php
declare(strict_types=1);

class Foo
{
    public function __construct(...$args)
    {
        var_dump($args);
    }
}

$foo = new Foo(a: 'a', b: 'b');

// ['a' => 'a', 'b' => 'b']
```

**Comment by brendt** [comment]
Though `func_get_args` should work according to the RFC. I'll check with internals.

**Comment by iamniels** [comment]
Hi Brendt, great work!

The last couple of weeks I've been thinking about a combination of a FormRequest and a DTO. For a form/ API call now I need to define the attributes twice, once in the FormRequest and once in the DTO. Would it be possible to combine these two classes without creating a mess? I imagine this would be a DTO which supports validation rules.

**Comment by JosephLeedy** [comment]
Hello Brendt, thank you for all of your hard work on this project! I agree with all of your points. Something that I'd definitely like to see added, as well, is support for the Spatie Enum package. I sometimes have APIs with fields that use specific values, so having an ENUM for them would help ensure the correct value is provided.

**Comment by jdart** [comment]
When talking about arrays, are you referring to lists, or key-value pairs, or both?

**Comment by aidan-casey** [comment]
@brendt - First of all, thank you for all the work you have put into this package. It has been very useful for our team once we fully got our heads around using it appropriately.

Now, I never would have said this when I first started using data transfer objects, but I'm just not sure that the features described above are within the scope of this package. DTOs are intended to contain / represent the data, but not manipulate it in this way. While there were some exceptions made to this in the past, they were minimal enough to not pull away from the original focus of data transfer objects.

It may be time to let this package die and instead put effort into tools that would compliment the idea of DTOs in PHP 8 classes. I would be happy to assist with these efforts if that's a route that is chosen. I could see two potential ideas right off the bat...

## Collections
While PHP 8 solves the problem of DTO classes, it does not necessarily solve how they are organized. Arrays are great, but collections are better. There could be room for a Spatie collection package which brings useful methods for interacting with multiple DTOs.

Validation might even be able to be implemented by defining which class the collection represents and confirming that with a reflection class. Example:

```php

use Example\SomeDTO;
use Spatie\Collection\Collection;

class SomeDTOCollection extends Collection
{
    private $collectionOf = SomeDTO::class;
}
```

I suppose there are plenty of other packages out there that provide collection classes, so it may not be necessary, but it is a value add that this package has traditionally carried with it.

## DTO Factory
To your point, casts have been something people have desired in the past. While they may not belong in _this_ package, they do have their place and it would seem to me that is within a factory.

```php

use Example\SomeDTO;
use Spatie\DTOFactory\Factory;

$dto = Factory::create(SomeDTO, [...]);

```

Factories can then be extendable to support things like creating a DTO from a Guzzle response.

```php

use GuzzleHttp\Psr7\Response;
use Spatie\DTOFactory\Factory;

class MyFactory extends Factory
{
    /**
     * Converts a Guzzle response to an array and then creates the DTO.
     */
    public static function createFromResponse(string $dto, Response $response)
    {
        $array = json_decode($response->getBody(), true);
        return static::create($dto, $array);
    }
}

```

## TLDR

PHP 8 covers the data transfer object piece, but there could be some opportunity to enhance that development experience with _other_ tools. That said, the potential problem with those other tools is that type casting may not be as great as driving the validation through the DTO class itself.

**Comment by ValiDrv** [comment]
I would add a few things:

#1 cache the reflection data the first time as object is used. That way, loops/long running process will speed up quite alot.

#2 for string/int public properties, add some extra metadata with some validation rules, example.

#[validate([a-z]+): error message]
#[validate(Validator:: CONSTANT)]
#[validate(Validator::custom_rule())]

#3 if you ever add arrays, add some validation rules, for one level deep at least.

/* @var array<into, string> */
#[validateKey([1-5]{32}): error message]
#[validateValue([a-z]+): error message]
#[validate(Validator::size(1,9)): error message]


It might be debated that the last two points are out of scope of the dto, but it makes the rest of the code much easier to work with.

**Comment by brendt** [comment]
> The last couple of weeks I've been thinking about a combination of a FormRequest and a DTO […] Would it be possible to combine these two classes without creating a mess? 

@iamniels handling form requests won't be be within the scope of this package. Having used DTOs all over the place for the past two years, I know that — at least in my projects — DTO type checks and request validation aren't the same. DTOs are about correctly typed data, while request validation encapsulates much more business logic.

> Something that I'd definitely like to see added, as well, is support for the Spatie Enum package

@JosephLeedy I agree, I'm not sure whether we could ship a built-in cast without a required dependency in this package or whether it should make some kind of DTO-enum-bridge package. I prefer to keep things simple though. I'd like to hear @Gummibeer his opinion as well.

> When talking about arrays, are you referring to lists, or key-value pairs, or both?

@jdart I'm talking about lists, everything else belongs in an object.

**Comment by brendt** [comment]
> Now, I never would have said this when I first started using data transfer objects, but I'm just not sure that the features described above are within the scope of this package.

@aidan-casey you're right saying that. Still I think it's better to continue with this package instead of creating a new one. This package is at ±750k downloads, and we don't want to bother our users with requiring a new package. I think the official description of v3 should be "a package that helps constructing DTOs in a pleasant manner". So I'd say `spatie/data-transfer-object` is still an accurate enough name.

> […] and it would seem to me that is within a factory.

I think it's a matter of coding style. While I appreciate the flexibility factories would offer, I don't think they require a package rewrite. I often use DTO factories in my project, which are simple classes that know about several ways of constructing a DTO. I don't see much functionality that could be abstracted to a package. 

Also keep in mind that lots of people are used to static constructors on the DTOs themselves, it's another style but equally valid. So I think it's important to not only focus on external factories, and keep the DTO-specific construction logic without the DTO itself.

**Comment by brendt** [comment]
> cache the reflection data the first time as object is used. That way, loops/long running process will speed up quite alot.

@ValiDrv This already happens: https://github.com/spatie/data-transfer-object/blob/master/src/DataTransferObject.php#L183 but I'll definitely also add it in the new version

> #2 for string/int public properties, add some extra metadata with some validation rules, example.

A few people have suggested this. Honestly I think it's a rabbit hole, because there are lots of possible validation rules, this package can't support all of them. What we could do is add some kind of generic validation attribute which you could implement with whatever business logic you'd want. I'll look into this!

> #3 if you ever add arrays, add some validation rules, for one level deep at least.

That's a great idea, but it also looks rather complex. Maybe custom validation attributes could help here.

**Comment by Gummibeer** [comment]
> Now, I never would have said this when I first started using data transfer objects, but I'm just not sure that the features described above are within the scope of this package.

> @aidan-casey you're right saying that. Still I think it's better to continue with this package instead of creating a new one. This package is at ±750k downloads, and we don't want to bother our users with requiring a new package. I think the official description of v3 should be "a package that helps constructing DTOs in a pleasant manner". So I'd say spatie/data-transfer-object is still an accurate enough name.

I agree with @aidan-casey and disagree with @brendt response.

Validating and casting is simply not the scope of a DTO. It's part of a validator, caster and should be DTO unaware/unspecific. Like Laravel Validator and custom Casts. The classes and interfaces are used by Form request and Eloquent Model but are not bound to them. You could for example simply use these classes to get DTOs validated & casted with the same classes and interfaces as everything else in your (Laravel) app.

Like Brent said in the initial post: casting was never in the scope of this package and this package was meant as a replacement for wanted core functionality until they are added - this time has come with PHP8. So the time should also have come to drop this package.
For the casting feature it should be evaluated if spatie wants to maintain a clone of Laravel casting for native PHP applications. Don't know if the need is there!?

For me this package has become fully obsolete and should be ended. Potentially lead to some smaller and focused packages doing casting, validating or similar.

But like @aidan-casey said:
> DTOs are intended to contain / represent the data, but not manipulate it in this way.

For me the focus of DTOs is the **T** (*transfer*). Lets move to real world - it's not job of the airplane to validate the passengers - it's the job of the airport security. The passengers allowed to enter the plane are just accepted by the airplane, moved to another place and leave out.

----

Regarding the enum scope - I don't see the reason for a new package as it's only `Enum::make($value)` and the Laravel bridge already offers transformers/casters for all common places.

**Comment by JosephLeedy** [comment]
@brendt Thanks for the response! With regards to the Enum package, I think that separate bridge library, as you suggested, would be the best way to avoid defining a hard dependency on it.

With regards to validation, while it would be a nice feature to have, I agree that it would be out of the scope of this package. It _would_, however, be nice if there was a separate package to accompany this one with that functionality. I'm currently using the Symfony validation component (stand-alone) in my projects.

I _do_ like the idea of casts as it would probably make working with data from complex APIs easier.

**Comment by Gummibeer** [comment]
@JosephLeedy what would you expect the Enum bridge to do? In fact the enum is instantiated by `Enum::make($value)` - am I right that you would want a method/cast that enforces an enum instance?
As the Laravel wrapper provides:
https://github.com/spatie/laravel-enum/blob/006fe3b00cf8ba82ec41ca4511bb2612b9867a12/src/Casts/Cast.php#L29-L49

I have 2 points:
1) a dedicated only for this method -too much effort for two lines of code.
2) it's tied to the decision if this package will remain with an adjusted scope of casting values.

The problem of self-validating/casting objects is that its logic runs every time I instantiate a DTO.
Let's say I have a request that is validated by Laravel/Symfony/... and I take part of the data and put it in a DTO, possibly a complex one with nested DTOs - this results in a useless `1+n` validation/casting cycle for all data.
Same for model attributes which should be validated before persistence.

Same for API responses - if you won't/can't trust the response it shouldn't be blindly thrown into a DTO and let's see if the DTO throws a `TypeError` or similar but it should be passed to a validator which throws correct and usable exceptions and only if the data passes the validator it should be passed to the DTO.

This COULD be done by a DTO in a before/setup like hook - but shouldn't be done by the DTO itself.

And to be honest - writing the next custom validator isn't cool as there are fully grown and widely accepted solutions, most frameworks provide there own.

--------

To solve all the possible expectations and so on in a clean way I have one idea.
@brendt what do you think about an abstract base class with some kind of event/hook system and nothing more.

```php
abstract class DTO
{
  public function __construct(...$args)
  {
    $this->booting();
    $this->assignValues(func_get_args());
    $this->booted();
  }

  protected function booting(): void
  {
    // like Laravel Model trait boot
    foreach($bootMethods as $method) {
      $this->$method();
    }
  }
}
```

This event system could be used by other packages adding a self-validation based on symfony or laravel validator, add casting system and so on.
And possibly the base class could/should implement interfaces like `Jsonable, ArrayAccess, Stringable` and so on.

This wouldn't add unused/controversial logic for all pure DTO users and allow for super powerful packages like `laravel-dto` and so on adding self-validation, casting and whatever they want.

**Comment by aidan-casey** [comment]
@Gummibeer - It seems like the only benefit to that approach would be its implementing of interfaces like `Jsonable`, `ArrayAccess`, and `Stringable` and is that so sought after that it is worth the package? If I am looking for that validation piece, why wouldn't I just create my own base DTO the contains that logic on construct?

It still feels like beating a dead horse to me.

**Comment by Gummibeer** [comment]
@aidan-casey sorry if that was wrongly phrased: I'm still and definitely on the "drop this package" side!
The last idea was only for the case this package will stay alive. One benefit of this would be something like a standard - like we did for the enum package.
It's not this much logic - but having a standard/interface makes things interchangeable.
And primary `ArrayAccess` and `Jsonable` isn't that easy as you have to collect all properties and it's a bunch of methods.

**Comment by brendt** [comment]
> by a DTO in a before/setup like hook

@Gummibeer this is a constructor's purpose, in my opinion. I see why you'd suggest the `booting` and `booted` methods, I'd consider it but I don't think they'd be actually used.

> Validating and casting is simply not the scope of a DTO

To be honest, nothing in this package has ever been in the scope of DTOs, but we had to call it something. All it ever did were these things:

- Extra runtime type checks based on doc blocks: not what DTOs are about
- A simplified form of collection classes: also not what DTOs are about
- A simplified form of named arguments: nope
- Automatic value casts: also not

In other words, this package has always been about two things: type checks and making constructing objects easier. We all agree that type checks aren't necessary anymore in PHP 8. So what's left? I agree that `data-transfer-object` is an unfortunate name. Maybe it should have been called simply "object" or "better-object" or "docblock-type-checker". I think I called it DTO because I was going to use this package with DTOs. But we've since also used it for events, resources and what not. 

So given that the package has never been about DTOs, I figure most people using wouldn't mind? I gather from the responses here and on Twitter that that's indeed the case. 

I'll ask around some more for people's opinion.

**Comment by brendt** [comment]
Let me add an example btw of what I think the new version would look like, maybe it would change anyone's mind 👍 

```php
class MyDTO extends DTO
{
    #[NumberBetween(1, 100)]
    public int $a;

    public OtherDTOCollection $collection;

    #[CastWith(ComplexObjectCaster::class)]
    public ComplexObject $complexObject;

    public ComplexObjectWithCast $complexObjectWithCast;

    public OtherDTO $otherDTO;
}

$dto = new MyDTO(
    a: 5,
    collection: [
        ['id' => 1],
        ['id' => 2],
        ['id' => 3],
    ],
    complexObject: [
        'name' => 'test',
    ],
    complexObjectWithCast: [
        'name' => 'test',
    ],
    otherDTO: ['id' => 5],
);
```

A few remarks:

- `NumberBetween` would be a user-implemented validation rule. This package wouldn't become a validation library, but would only offer hooks to be able to write validators yourself.
- Both named arguments and arrays can be used to construct a DTO
- DTOs and DTOCollections are casted automatically
- `#[CastWith]` can be used directly on a DTO property, but also on the object's class itself:

```php
#[CastWith(ComplexObjectWithCastCaster)]
class ComplexObjectWithCast
{
    // …
}
```

**Comment by mfn** [comment]
> `#[CastWith]` can be used directly on a DTO property, but also on the object's class itself:

What be nice to have a way to define this for classes we don't control; think about `DateTimeImmutable` or `Carbon\CarbonImmutable`, etc so I don't have to specify this everywhere I use such an attribute.

A property `CastWith` with still be authoritative.

**Comment by brendt** [comment]
> What be nice to have a way to define this for classes we don't control; think about DateTimeImmutable or Carbon\CarbonImmutable, etc so I don't have to specify this everywhere I use such an attribute.

That's a good idea, I'll think about an implementation.

**Comment by Gummibeer** [comment]
@brendt how about an attribute on a base DTO class like `#[CastUsing(DateTime::class, DateTimeCaster::class)]`? 🤔

And regarding the scope of this package - I think that a real list of features/classes/interfaces would help a lot!
From what I see in your example code I would be fine. A feature I would like is any kind of flag that will simply skip the custom behavior. In best case a runtime flag so that I can do something like `MyDto::make($raw)` (this will go through all the casting) and `MyDto::fill($data)` (which will just put everything on the properties).
Do you understand what I mean?^^

**Comment by brendt** [comment]
> Do you understand what I mean?^^

I think I do. Are there use cases where you want both behaviours on the same class? I figure you'd opt for not extending from the DTO class if you don't want the behaviour?

**Comment by brendt** [comment]
> I think that a real list of features/classes/interfaces would help a lot!

I'll work on an updated version of the README and send a PR, that's the best way to visualize the changes I believe.

**Comment by Gummibeer** [comment]
I think about cases like:

```php
$dto1 = Dto1::make($data1);
$dto2 = Dto2::fill([
  'foo' => 'bar',
  'dto_1' => $dto1,
]);
```

This code is super reduced - but I have several cases when data comes from different sources, through different pipelines but ends in the same DTO. An `Address` DTO for example. If I fill it from my DB I don't have to validate/cast it anymore, same if I get the data from a request which is already validated.
But if I get the data from an API, Here or Google Maps, for example I would like to pipe them through the casting as I'm not sure what they send my way. Same for mass imports from a CSV our company customer has sent me.

So for me there are two points:
* How important is the performance for this instantiation?
* How trustworthy is the source of the raw data?

**Comment by brendt** [comment]
FYI here are some examples: https://github.com/spatie/data-transfer-object/blob/v3/README.md

**Comment by ValiDrv** [comment]
I have a few suggestions (based on a few modifications we had in-house):

#1 Should have a way to validate collections also, for those simple list ones.

```
    /* array<int, string> */
    #[length(2, 4)]            --> validates entire array
    #[value:regex('[a-z]+')]   --> validates each value, one by one.
    #[key:regex('\d+')]        --> validates each key, one by one.
    public array $collection;  --> might have to be OtherDTOCollection instead of array, due to how PHP works...
```

#2 Maybe to simplify things, why not let the validation function return `true`/`false`, and throw an `exception` if it needs to send a custom message? That way, you can use normal PHP functions in there also, which throw exceptions if they failed. The constructor could buffer all of those, and re-throw an exception with all the fields that failed.
```
public function validate(mixed $value): bool
    {
        if ($value < $this->min) {
            throw new InvalidArgumentException("Value should be greater than or equal to {$this->min}");
        }
        ...
        return true;
    }
```

#3 The validation function should probably take the class and property it's validating.
This can be useful in case of custom error messages/if people want to use it to validate forms.
```
    public function validate(mixed $value, string $className, string $propertyName): ValidationResult
```

#4 Some async libraries like `swoole` have some limitations caused by php core code. 
They cannot have async code in methods like `__get`/`__set` (see: [google translate to english this URL](https://wiki.swoole.com/#/coroutine/notice?id=%E5%9C%A8-__get-__set-%E9%AD%94%E6%9C%AF%E6%96%B9%E6%B3%95%E9%87%8C%E4%B8%8D%E5%BE%97%E4%BA%A7%E7%94%9F%E5%8D%8F%E7%A8%8B%E5%88%87%E6%8D%A2) )

This can cause issues with validations that need some IO (db/file system access), and you can go around it with a method call or `__call`. So there should be a `$obj->set('key','value')`


#5 need a way to serialize/deserialize the object (and all it's children) any way we want. (the php `serialize` is kind of slow)
```
$data = json_encode($obj->dump()); 
$obj = new Obj( json_decode( $data ,true) );
```
_Note:_ for the "all it's children" part, If we have objects referencing other objects, this can get tricky, but a simple depth level check could be enough.

**Comment by jessarcher** [comment]
> For the casting feature it should be evaluated if spatie wants to maintain a clone of Laravel casting for native PHP applications. Don't know if the need is there!?

I would still appreciate the casting functionality, even within a Laravel application, as I don't think it's really a clone. I think it still has value when dealing with JSON columns, particularly with nested objects.

I recently created https://github.com/jessarcher/laravel-castable-data-transfer-object which combines this package with Laravel's Castable interface to automatically cast between a JSON column and a custom value object.

Of course if Spatie doesn't want to continue maintaining this I completely understand and will consider pulling any necessary functionality over into my package.

**Comment by DanielSchetritt** [comment]
I would love the ability to be able to use a callback function for casting.

For example, if I want to cast an assoc array that represents a fixed-point decimal number as a php float type:
```php
// Defined the callback somewhere
function ($value) {
    ['exponent' => $exponent, 'mantissa' => $mantissa] = $value;
    return floatval("{$mantissa}e{$exponent}");
}
// This is used in creating the DTO
$dto = new SomeDto([
    'fixed_decimal' => [
        'exponent' => -2,
        'mantissa' => 123,
    ],
]);
$true = $dto->fixed_decimal === 1.23
```

**Comment by aidan-casey** [comment]
I was thinking about the DTO collections the other day. I am wondering if it may be a good idea to pull an idea out of the book of the [Ramsey Collection](https://github.com/ramsey/collection) package. When constructing a collection, the first parameter has to be a string representation of the FQDN of the class the collection is for. This allows it to do type validation on anything added to the collection.

Example:

```php

use DTOs\Car;

// Creates an empty collection that represents cars
$cars = new Collection(Car::class, []);

```

Since the problem this package concerns type safety, it would be nice to see that functionality on the collection class and really not that hard to add support for.

That said, this isn't another collection package, so how far we develop those collections becomes a question in my mind.

**Comment by aidan-casey** [comment]
Also @brendt, any plans to dump the chosen features into a project or issues with milestones so we can assist as ideas are firmed up?

**Comment by brendt** [comment]
> When constructing a collection, the first parameter has to be a string representation of the FQDN of the class the collection is for. This allows it to do type validation on anything added to the collection.

It's somewhat a replacement for a problem that generics would solve. The issue with it is that there's only runtime type validation. Static analysers won't be able to tell what's in the collection. Especially with the focus shift towards static analysis in v3, I don't think it would add much  value.

> any plans to dump the chosen features into a project or issues with milestones so we can assist as ideas are firmed up?

Yes, I'll work on this the coming weeks. Feel free to comment on https://github.com/spatie/data-transfer-object/pull/159 and send PRs to the v3 branch.

**Comment by wedi** [comment]
@brendt I like the direction of v3!
I found your PR draft https://github.com/spatie/data-transfer-object/pull/159/ and was wondering what the status of this early version is. Could you give a short update?

**Comment by brendt** [comment]
It's on my todo list but there are a few other packages I'm working on first 😅 so don't expect this to land in the coming weeks. But I haven't forgotten about it!

**Comment by vanthao03596** [comment]
Can we add feature to map DTOA to DTOB with convention key from camelcase to snakekey like [nartc/mapper](https://automapperts.netlify.app/docs/getting-started/introduce-to-automapper).

**Comment by olivernybroe** [comment]
Hi, just wanted to share what our company did as our solution for DTO's with PHP8, we have been using it for about 3 months and are pretty happy with it.

We use named arguments and property promotion.
Biggest waste is prob. our usage of accessors.

We then have static methods for easy building of the DTO from a request, as we use GQL also, we sometimes also have a `fromGraphQl` method also.

<details>
<summary>DTO example</summary>

```php
class BillingInformationDTO
{
    #[Pure]
    public function __construct(
        private bool $enabled,
        private LocationDTO $billingLocation,
        private Company $company,
        private ?string $companyName = null
    )
    { }

    #[Pure]
    public static function fromRequest(UpdateProfileRequest $request, Company $company): self
    {
        $billingLocation = new LocationDTO(
            address: $request->get('billing_address'),
            zipcode: $request->get('billing_zipcode'),
            city: $request->get('billing_city'),
            state: $request->get('billing_state'),
            country: $request->get('billing_country'),
            province: $request->get('billing_province')
        );
        return new self(
            enabled: !$request->get('disable_billing_information', false),
            billingLocation: $billingLocation,
            company: $company,
            companyName: $request->get('billing_company_name')
        );
    }

    public function isEnabled(): bool
    {
        return $this->enabled;
    }

    public function getCompanyName(): ?string
    {
        return $this->companyName;
    }

    public function getCompany(): Company
    {
        return $this->company;
    }

    public function getBillingLocation(): LocationDTO
    {
        return $this->billingLocation;
    }
}
```

</details>

<details>
<summary>DTO usage</summary>

```php
$billingInformationDTO = BillingInformationDTO::fromUpdateProfileRequest($request, Auth::company());

$updateBillingInformation->fromDTO($billingInformationDTO)
	->update();
```

</details>

**Comment by Gummibeer** [comment]
@olivernybroe you could replace all these accessors with public properties and `@psalm-readonly` annotation if you use psalm.
Alternatively you can use a `__call()` or `__get()` magic method to automatically access all the protected properties.

**Comment by Gummibeer** [comment]
Yeah, the first works only if you use a static analyzer. Which will be even better than PHP exploding.
But was just feedback - I think that the magic way would be the only relevant for this package - if it proceeds. But as you showed - DTOs with PHP8 are that simple that this package would be more about casting and other features than DTOs.

**Comment by brendt** [comment]
FYI I added the last pending feature: the ability to configure default casts on a DTO level. Here's the PR: https://github.com/spatie/data-transfer-object/pull/159

**Comment by brendt** [comment]
We'll discuss all furthers things in the PR
